cmd/qat_plugin/dpdkdrv/dpdkdrv.go

```diff
@@ -245,56 +245,108 @@ func (dp *DevicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
 	return nil
 }
 
+func getPciDevicesWithPattern(pattern string) (pciDevices []string) {
+	pciDevices = make([]string, 0)
+
+	devs, err := filepath.Glob(pattern)
+	if err != nil {
+		klog.Warningf("bad pattern: %s", pattern)
+		return
+	}
+
+	for _, devBdf := range devs {
+		targetDev, err := filepath.EvalSymlinks(devBdf)
+		if err != nil {
+			klog.Warningf("unable to evaluate symlink: %s", devBdf)
+			continue
+		}
+		pciDevices = append(pciDevices, targetDev)
+	}
+
+	return
+}
+
+func (dp *DevicePlugin) getVfDevices() []string {
+	qatPfDevices := make([]string, 0)
+	qatVfDevices := make([]string, 0)
+
+	// Get PF BDFs bound to a PF driver
+	for _, vfDriver := range dp.kernelVfDrivers {
+		pfDriver := strings.TrimSuffix(vfDriver, "vf")
+		pattern := filepath.Join(dp.pciDriverDir, pfDriver, "????:??:??.?")
+		qatPfDevices = append(qatPfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	// Get VF devices belonging to a PF device
+	for _, qatPfDevice := range qatPfDevices {
+		pattern := filepath.Join(qatPfDevice, "virtfn*")
+		qatVfDevices = append(qatVfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	if len(qatVfDevices) > 0 {
+		return qatVfDevices
+	}
+
+	// No PF devices found, running in a VM?
+	for _, vfDriver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
+		pattern := filepath.Join(dp.pciDriverDir, vfDriver, "????:??:??.?")
+		qatVfDevices = append(qatVfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	return qatVfDevices
+}
+
+func getCurrentDriver(device string) string {
+	symlink := filepath.Join(device, "driver")
+	driver, err := filepath.EvalSymlinks(symlink)
+	if err != nil {
+		klog.Warningf("unable to evaluate symlink: %s", symlink)
+		return ""
+	}
+	return filepath.Base(driver)
+}
+
 func (dp *DevicePlugin) scan() (dpapi.DeviceTree, error) {
 	devTree := dpapi.NewDeviceTree()
 	n := 0
-	for _, driver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
-		files, err := ioutil.ReadDir(filepath.Join(dp.pciDriverDir, driver))
+
+	for _, vfDevice := range dp.getVfDevices() {
+		vfBdf := filepath.Base(vfDevice)
+
+		vfdevID, err := dp.getDeviceID(vfBdf)
 		if err != nil {
-			klog.Warningf("Can't read sysfs for driver as Driver %s is not available: Skipping", driver)
+			return nil, errors.Wrapf(err, "Cannot obtain device ID for %s", vfBdf)
```

Comment:
nit: no need to embed one more stack trace into the error. Can be fixed in a followup PR though as the existing code in the master branch does the same.